### ISSUE TITLE
Common Interfaces

### ISSUE THREAD

**Issue by PleasantD** [jsgoupil/quickbooks-sync#44]
It would be very helpful if common types had either common base types or common interfaces.

For Example, `VendorAddress` and `BillAddress` are both identical address types.
But since the generated code doesn't generate them from a common type or assign a common interface, mapping code requires either reflection or copy-paste duplication.

Another example would be the common properties of all Mod types or Ret types (`ListId`, `EditSequence`, etc.).

If the XSD and generator can't be updated to do this automatically, it would be possible to add them manually if the generator built all types with the `partial` keyword. Then common interfaces could be manually added to your codebase and associated with appropriate types.

**Comment by tofer** [comment]
I have a fork that includes some common interfaces for the Ret types. I don't think I did anything with `VendorAddress` and `BillAddress`, but on quick check both of those are already partials. I have been meaning to do a PR for a long time. I'll try to go look at the status of my fork and see if I can get that in.

@jsgoupil accepts PRs if you wanted to add the interface for `VendorAddress` and `BillAddress` and any of the address address types that have common fields.

**Comment by jsgoupil** [comment]
Adding a `partial` is a good idea, however, it also allows you to add fields/properties. And that would be problematic for the XML generation. This would allow people to add types the way they want.
Adding an interface ourselves is also a possibility but the maintainability can be quite the nightmare; finding common types can be quite suggestive. Deciding what is a common type ourselves can be complicated.

What do you think @tofer regarding this? what approach would you go? I don't see anything being partial right now.

**Comment by tofer** [comment]
OK, PR #45 created

**Comment by tofer** [comment]
@jsgoupil You don't see anything partial? Seems like every class generated in `Objects.cs` is partial. 

Looking at the XSD, we should absolutely be able to add a common interfaceto the Addresses. Most of them reference either group `AddressData` or `AddressBlockData`. Might be able to do it similarly to the way I did for the *Ret intefaces I just did the PR for.

```xml
<xsd:element name="LegalAddress">
	<xsd:complexType>
		<xsd:sequence>
			<xsd:group ref="AddressData"/>
		</xsd:sequence>
	</xsd:complexType>
</xsd:element>

<xsd:element name="BillAddress">
	<xsd:complexType>
		<xsd:sequence>
			<xsd:group ref="AddressData"/>
		</xsd:sequence>
	</xsd:complexType>
</xsd:element>

<xsd:element name="BillAddressBlock">
	<xsd:complexType>
		<xsd:sequence>
			<xsd:group ref="AddressBlockData"/>
		</xsd:sequence>
	</xsd:complexType>
</xsd:element>

<xsd:element name="VendorAddress">
	<xsd:complexType>
		<xsd:sequence>
			<xsd:group ref="AddressData"/>
		</xsd:sequence>
	</xsd:complexType>
</xsd:element>
```

**Comment by tofer** [comment]
Well it was _really_ easy to add interfaces for the `AddressData` and `AddressBlockData` groups in the XSD (`IQbAddress` and `IQbAddressBlock`), so I went ahead and just did it and pushed another commit to the PR [here](https://github.com/jsgoupil/quickbooks-sync/pull/45/commits/9fdd99ba5269ec52d4b91f4bb734deb9c4b2cc2b).

While I was in there, I noticed some other very commonly used groups, namely `ContactInfo`, `CommInfo`, and `PersonName`, so I also added interfaces `IQbContactInfo`, `IQbCommInfo`, and `IQbPersonName` respectively as well in case that is useful.

**Comment by jsgoupil** [comment]
Strange on the partial...
![image](https://user-images.githubusercontent.com/808649/83806624-aa2a5f00-a666-11ea-82b7-1d65530549bb.png)

Maybe it has to do with how it's packaged?

**Comment by tofer** [comment]
Oh yea, I forgot partials all have to be in the same library. So if you are decompiling to view source, that is probably not shown since it can't be treated as partial across libraries.

![image](https://user-images.githubusercontent.com/1517583/83807810-81a36480-a668-11ea-9a41-b21a843b120f.png)

**Comment by jsgoupil** [comment]
So he couldn't really do partial then in his own project. We needed a PR.
Thanks for that PR. I will be updating this real soon.

**Comment by PleasantD** [comment]
> So he couldn't really do partial then in his own project. We needed a PR.

This is correct. Partials only exist withing a compilation unit and don't allow for extending a type.
They are just a friendly way for a large type to span multiple files, or for generated code to allow for custom extension _within_ the compilation unit it is generated in.

This kind of common interfacing will help reduce the duplication I currently need.
When is a new release expected?

**Comment by jsgoupil** [comment]
@PleasantD Between now and Sunday.

**Comment by jsgoupil** [comment]
Fixed in 2644a55d928f58769b38b09fdda03983708ae691
Released as qbxml-2.2.0
